blog/2022/04/simplechains.md

```diff
@@ -0,0 +1,489 @@
+2. Focus on accelearting GPU kernels as fast as possible! Since these large matrix-matrix operations will be fastest on GPUs and are the bottleneck, performance benchmarks will essentially just be a measurement of how fast these specific kernels are.
+3. When doing reverse-mode automatic differentiation (backpropogation), feel free to copy values to memory. Memory allocations will be hidden by the larger kernel calls. 
+4. Also, feel free to write a "tape" for generation backpropogation. The tape does add the cost of essentially building a dictionary during the forward pass, but that will be hidden by the larger kernel calls.
+
+Do these assumptions actually hold in our case? And if they don't, can we focus on these aspects to draw more performance out for our use cases?
+
+## Digging In: Small Neural Network Performance Overheads
+
+It's easy to show that these assumptions breakdown when we start focusing on this smaller neural network use case. For starters, lets look at assumptions (1)(2). It's not hard to show where these two will unravel:
+
+```julia
+using LinearAlgebra, BenchmarkTools, CUDA, LoopVectorization
+
+function mygemmavx!(C, A, B)
+    @tturbo for m ∈ axes(A, 1), n ∈ axes(B, 2)
+        Cmn = zero(eltype(C))
+        for k ∈ axes(A, 2)
+            Cmn += A[m, k] * B[k, n]
+        end
+        C[m, n] = Cmn
+    end
+end
+
+function alloc_timer(n)
+    A = rand(Float32,n,n)
+    B = rand(Float32,n,n)
+    C = rand(Float32,n,n)
+    t1 = @belapsed $A * $B
+    t2 = @belapsed (mul!($C,$A,$B))
+    t3 = @belapsed (mygemmavx!($C,$A,$B))
+    A,B,C = (cu(A), cu(B), cu(C))
+    t4 = @belapsed CUDA.@sync($A * $B)
+    t5 = @belapsed CUDA.@sync(mul!($C,$A,$B))
+    t1,t2,t3,t4,t5
+end
+ns = 2 .^ (2:11)
+res = [alloc_timer(n) for n in ns]
+alloc      = [x[1] for x in res]
+noalloc    = [x[2] for x in res]
+noalloclv  = [x[3] for x in res]
+allocgpu   = [x[4] for x in res]
+noallocgpu = [x[5] for x in res]
+
+using Plots
+plot(ns, alloc, label="*", xscale=:log10, yscale=:log10, legend=:bottomright,
+    title="Which Micro-optimizations matter for BLAS3?",
+    yticks=10.0 .^ (-8:0.5:2),
+    ylabel="Time (s)", xlabel="N",)
+plot!(ns,noalloc,label="mul! (OpenBLAS)")
+plot!(ns,noalloclv,label="mygemmavx!")
+plot!(ns,allocgpu,label="* gpu")
+plot!(ns,noallocgpu,label="mul! gpu")
+savefig("microopts_blas3.png")
+```
+
+![](https://user-images.githubusercontent.com/1814174/162625573-0c195a60-dc06-44fb-af04-6c0c389ded02.png)
+
+When we get to larger matrix-matrix operations, such as 100x100 * 100x100, we can effectively write off any overheads due to memory allocations. But we definitely see that there is a potential for some fairly significant performance gains in the lower end! Notice too that these gains are realized by using the pure-Julia LoopVectorization.jl as the standard BLAS tools tend to have extra threading overhead in this region (again, not optimizing as much in this region). 
```

Comment:
Can we?
What if we ran matmuls in multiple threads, and took the mean time instead of the minimum?
Mean is much more representative of actual performance whenever the GC is involved.

But if we don't want to make an overly complicated benchmark just for the sake of demoing how our GC doesn't scale with threads, maybe at least using `mean` instead of `minimum` times in our benchmarks will be more accurate, while also helping get this point across?

The easiest way to get mean times out of BenchmarkTools I'm aware of:
```julia
mean(@benchmark rand(40)).time
```

Reply:
Yeah that's probably worth trying.